File: backend/src/sql/ai.ts
export const SQL_SELECT_EMBEDDINGS_BY_KEYS = `
  SELECT cache_key, embedding
  FROM embedding_cache
  WHERE cache_key = ANY($1::text[])
`;

export const SQL_UPSERT_EMBEDDING_CACHE = `
  INSERT INTO embedding_cache (cache_key, provider, model, dimensions, embedding, updated_at)
  VALUES ($1, $2, $3, $4, $5::jsonb, now())
  ON CONFLICT (cache_key) DO UPDATE SET
    embedding = EXCLUDED.embedding,
    updated_at = now()
`;

File: apps/api/app/sql/ai.py
from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.dialects.postgresql import array, insert as pg_insert

from app.db.tables import embedding_cache


def select_embeddings_by_keys(keys: list[str]):
    return select(embedding_cache.c.cache_key, embedding_cache.c.embedding).where(
        embedding_cache.c.cache_key.in_(keys)
    )


def upsert_embedding_cache(
    cache_key: str,
    provider: str,
    model: str,
    dimensions: int,
    embedding: list,
):
    stmt = pg_insert(embedding_cache).values(
        cache_key=cache_key,
        provider=provider,
        model=model,
        dimensions=dimensions,
        embedding=embedding,
        updated_at=func.now(),
    )
    return stmt.on_conflict_do_update(
        index_elements=[embedding_cache.c.cache_key],
        set_={
            "embedding": stmt.excluded.embedding,
            "updated_at": func.now(),
        },
    )
